qt/aqt/deckoptions.py

```diff
@@ -57,15 +60,60 @@ def _setup_ui(self) -> None:
 
     def _on_bridge_cmd(self, cmd: str) -> None:
         if cmd == "deckOptionsReady":
+            self.ready = True
             gui_hooks.deck_options_did_load(self)
 
+    def closeEvent(self, evt: QCloseEvent) -> None:
+        if self._close_event_has_cleaned_up:
+            evt.accept()
+            return
+        evt.ignore()
+        self.if_can_close()
+
+    def _close(self):
+        """Close. Ensure the closeEvent is not ignored."""
+        self._close_event_has_cleaned_up = True
+        self.close()
+
+    def if_can_close(self):
+        """Close if there was no modification. Otherwise ask for confirmation first."""
+
+        def callbackWithUserChoice(choice: int):
+            if choice == 0:
+                # The user accepted to discard current input.
+                self._close()
+
+        def if_can_close_callback_with_data_information(has_modified_dataData: bool):
+            if has_modified_dataData:
+                ask_user_dialog(
+                    tr.card_templates_discard_changes(),
+                    callback=callbackWithUserChoice,
+                    buttons=[
+                        QMessageBox.StandardButton.Discard,
+                        (tr.adding_keep_editing(), QMessageBox.ButtonRole.RejectRole),
+                    ],
+                )
+            else:
+                self._close()
+
+        self.has_modified_data(if_can_close_callback_with_data_information)
+
     def reject(self) -> None:
         self.mw.col.set_wants_abort()
         self.web.cleanup()
         self.web = None
         saveGeom(self, self.TITLE)
         QDialog.reject(self)
 
+    def has_modified_data(self, callback: Callable[[bool], None]):
+        """Calls `callback` with the information of whether any deck options are modified."""
+        if self.ready:
+            self.web.evalWithCallback(
+                "globalThis.anki.deckOptionsPendingChanges()", callback
```

Comment:
```suggestion
                "anki.deckOptionsPendingChanges()", callback
```

globalThis is useful for bypassing typechecking; we don't gain anything by using it in the Python code